examples/bpi-monolith-test-suite/bpi/components/vsm/vsm.ts

```diff
@@ -0,0 +1,43 @@
+import { Agreement } from "../storage/agreement";
+import { Transaction } from "../transactions/transaction";
+import { IStorageComponent } from "../storage/storage.interface";
+import { IWorkgroupComponent } from "../workgroup/workgroup.interface";
+import { IVirtualStateMachineComponent } from "./vsm.interface";
+
+
+export class VirtualStateMachine implements IVirtualStateMachineComponent {
+    storageComponent: IStorageComponent;
+    workgroupComponent: IWorkgroupComponent;
+
+    constructor(storageComponent: IStorageComponent, workgroupComponent: IWorkgroupComponent) {
+        this.storageComponent = storageComponent;
+        this.workgroupComponent = workgroupComponent;
+    }
+
+    executeTransaction(transaction: Transaction): string {
+        // TODO: Poor man's deep copy - change
+        const latestAgrementState = this.storageComponent.getAgreementState();
+        const originalAgreementState = JSON.parse(JSON.stringify(latestAgrementState));
+
+        const workgroup = this.workgroupComponent.getWorkgroupById(transaction.workgroupId);
+        const workstep = workgroup.getWorkstepById(transaction.workstepId);
+
+        // If succesfull, this changes the agreement state
+        const workstepResult = workstep.execute(latestAgrementState, transaction.stateObject);
+
+        if (!workstepResult) {
+            return "err: workstep execution failed to satisfy the agreement.";
+        }
+
+        // TODO: Should this be in the workstep execution?
```

Comment:
yes it should be in `workstep.execute` @ognjenkurtic

Reply:
ok, noted